File: tools/person-cutout.m
#import <AppKit/AppKit.h>
#import <CoreImage/CoreImage.h>
#import <Vision/Vision.h>

int main(int argc, const char *argv[]) {
    @autoreleasepool {
        if (argc != 3) {
            fprintf(stderr, "Usage: person-cutout input output\n");
            return 2;
        }

        NSURL *inputURL = [NSURL fileURLWithPath:[NSString stringWithUTF8String:argv[1]]];
        NSURL *outputURL = [NSURL fileURLWithPath:[NSString stringWithUTF8String:argv[2]]];
        CIImage *source = [CIImage imageWithContentsOfURL:inputURL];
        if (!source) return 3;

        VNGeneratePersonSegmentationRequest *request = [VNGeneratePersonSegmentationRequest new];
        request.qualityLevel = VNGeneratePersonSegmentationRequestQualityLevelAccurate;
        request.outputPixelFormat = kCVPixelFormatType_OneComponent8;

        VNImageRequestHandler *handler = [[VNImageRequestHandler alloc] initWithCIImage:source options:@{}];
        NSError *error = nil;
        if (![handler performRequests:@[request] error:&error]) {
            fprintf(stderr, "%s\n", error.localizedDescription.UTF8String);
            return 4;
        }

        VNPixelBufferObservation *observation = request.results.firstObject;
        if (!observation) return 5;

        CIImage *mask = [CIImage imageWithCVPixelBuffer:observation.pixelBuffer];
        CGRect extent = source.extent;
        CGFloat sx = CGRectGetWidth(extent) / CGRectGetWidth(mask.extent);
        CGFloat sy = CGRectGetHeight(extent) / CGRectGetHeight(mask.extent);
        mask = [[mask imageByApplyingTransform:CGAffineTransformMakeScale(sx, sy)] imageByCroppingToRect:extent];

        CIImage *clear = [[CIImage imageWithColor:[CIColor clearColor]] imageByCroppingToRect:extent];
        CIFilter *blend = [CIFilter filterWithName:@"CIBlendWithMask"];
        [blend setValue:source forKey:kCIInputImageKey];
        [blend setValue:clear forKey:kCIInputBackgroundImageKey];
        [blend setValue:mask forKey:kCIInputMaskImageKey];

        CIContext *context = [CIContext contextWithOptions:@{}];
        CGColorSpaceRef colorSpace = CGColorSpaceCreateWithName(kCGColorSpaceSRGB);
        NSData *png = [context PNGRepresentationOfImage:blend.outputImage
                                                 format:kCIFormatRGBA8
                                             colorSpace:colorSpace
                                                options:@{}];
        CGColorSpaceRelease(colorSpace);
        if (!png || ![png writeToURL:outputURL options:NSDataWritingAtomic error:&error]) {
            fprintf(stderr, "%s\n", error.localizedDescription.UTF8String);
            return 6;
        }
    }
    return 0;
}

File: tools/person-cutout.swift
import AppKit
import CoreImage
import Vision

guard CommandLine.arguments.count == 3 else {
    fputs("Usage: person-cutout input output\n", stderr)
    exit(2)
}

let inputURL = URL(fileURLWithPath: CommandLine.arguments[1])
let outputURL = URL(fileURLWithPath: CommandLine.arguments[2])

guard let source = CIImage(contentsOf: inputURL) else {
    fputs("Could not read input image\n", stderr)
    exit(3)
}

let request = VNGeneratePersonSegmentationRequest()
request.qualityLevel = .accurate
request.outputPixelFormat = kCVPixelFormatType_OneComponent8

let handler = VNImageRequestHandler(ciImage: source, options: [:])
try handler.perform([request])

guard let observation = request.results?.first else {
    fputs("No person mask generated\n", stderr)
    exit(4)
}

let mask = CIImage(cvPixelBuffer: observation.pixelBuffer)
let scaleX = source.extent.width / mask.extent.width
let scaleY = source.extent.height / mask.extent.height
let scaledMask = mask.transformed(by: CGAffineTransform(scaleX: scaleX, y: scaleY))
    .cropped(to: source.extent)

let clear = CIImage(color: .clear).cropped(to: source.extent)
let output = source.applyingFilter(
    "CIBlendWithMask",
    parameters: [
        kCIInputBackgroundImageKey: clear,
        kCIInputMaskImageKey: scaledMask
    ]
)

let context = CIContext(options: [.useSoftwareRenderer: false])
let colorSpace = CGColorSpace(name: CGColorSpace.sRGB)!
guard let data = context.pngRepresentation(
    of: output,
    format: .RGBA8,
    colorSpace: colorSpace
) else {
    fputs("Could not encode output PNG\n", stderr)
    exit(5)
}
try data.write(to: outputURL)
